### PR TITLE
Add SimpleJWT authentication support, cleanups

### PR DESCRIPTION
**Pull request by dmaone**
* Transparent support for Django SimpleJWT authentication variant
* Tests for SimpleJWT workflow added
* Fixed scenario where base URL not ending with slash will create invalid URL paths
* unittest2 is for python2 - py3 has `unittest` in standard library
* pytest-localserver is dead since 2017 and had a hardcoded certificate which no longer acceptable by OpenSSL, failing SSL tests
* Removed unused `_handle_redirect` from RestResource
* Removed unneeded `_copy_kwargs` and `_iterator` methods - there's `dict.copy()` since py2.2
* Enforcing consistent use of `requests.Session` in API
* Using that session object to store common headers - including auth.
* de-generalized self._store - `session` and `base_url` aren't really
optional in our case.
* `requests` knows how to handle JSON - no need for dumps/loads
* centralized SSL exception conversion
* `black` formatter settings

### CODE REVIEW

File: archfx_cloud/api/connection.py
Comment (dkarchmer) [inline archfx_cloud/api/connection.py:293]: This is technically not needed as `auth/api-jwt-refresh` is a custom method, so we can take whatever we want and convert to what the library requires on the cloud/broker side.

But the real issue here is that during migration, we may need to deploy two APIs side by side, so we may need to send to a different end point depending on the package. But that will only be for a few days so maybe we just ignore that.

File: archfx_cloud/api/connection.py
Comment (dkarchmer) [inline archfx_cloud/api/connection.py:236]: You are making an assumption here that we will want to return the data as 
```
"jwt": {
      "access"...
      "refresh"....
}
```

which does not necessary need to be true. The fact is that if we later worry that we want to be backwards compatible, we will probably instead return

```
"jwt": "access-key",
"refresh-jwt": "refresh-key"
```

I like your proposal better, but it does break any user out there (not using our package). We do have at least two customers using our API today (not likely using "refresh" but for sure using the "jwt" token

**Review by dkarchmer — COMMENTED** [review]
Thanks for the cleanup. This library pre-dates Arch (do you want to update https://github.com/dkarchmer/django-rest-framework-client as well?  Just kidding).

But before I can approve, we need to finalize our migration plan. That's important as depending how backwards compatible we want to be, your payload assumptions may be wrong. Our cloud/broker methods are customizable, so we can be fairly backwards compatible (except from the new requirement for a special "refresh" token, but most users don't use that).

My worry is that we do have at least two customers using the API (and for sure one without out python package). I do like using the `"jwt": Dict` option, but I will need to be sure that I can reach them and have them migrate too.

So, the cleanup looks great, but the code to make SimpleJWT work transparently may or may not be correct. That said, I may be ok approving for now, and changing later if we believe we need to be more backwards compatible

File: archfx_cloud/api/connection.py
Comment (dmaone) [reply archfx_cloud/api/connection.py:293]: The main idea is this client implementation supports both old and new REST JWT styles, transparently to the user.

The central cloud instance can remain on `djangorestframework-jwt==1.11.0` until you want to upgrade to Django 3.2 (`djangorestframework-jwt` pins `djangorestframework` to version 3.11, (which already doesn't support Django 3.1, but still works, warning about that it will break on Django 3.2) - giving us an almost infinite migration window.

`refresh_token_data` being there means that `SimpleJWT` is used (because `SimpleJWT` returns `access` and `refresh` fields in payload instead of `token` and that's how module learns what kind of server it's talking to).
Since "refresh token" views are part of JWT plugin and expect different fields to be in JSON - client asks for a refresh in a format the server will understand.

Re: customers using the API - are we talking about on-premise servers or the cloud server?
Because I just found 4 different implementations of ArchFX cloud client in `archfx_broker` repo and going to unify those.

Re:  https://github.com/dkarchmer/django-rest-framework-client - is it still used anywhere?

File: archfx_cloud/api/connection.py
Comment (dmaone) [reply archfx_cloud/api/connection.py:236]: The difference goes farther than that. SimpleJWT requires using separate refresh token to refresh access token (this is more secure - probably JWT spec was changed to require this since the old JWT plugin was released). So attempt to refresh the token using SimpleJWT-generated access token will fail.

Another thing is that `api-jwt-refresh/` view will return the same `{"access": "x", "refresh: "y"}` payload - it's easier to just use the same format.

Comment (dmaone) [reply archfx_cloud/api/connection.py:236]: ..but since not many people use the refresh (if any), and I would guess most people use `auth/login` instead of `auth-jwt-login` - we can have best of both worlds. "jwt" will still be a string, but presense of "jwt-refresh" would signal SimpleJWT and different handling of the refresh it requires.

File: archfx_cloud/api/connection.py
Comment (dkarchmer) [reply archfx_cloud/api/connection.py:293]: Yes, as I said, I am ok with this changes just to unblock the broker, but I also don't see the value of cleaning up the broker and not the cloud. All I am saying is that we are not done until everything is upgrading, so upgrading the broker is actually fairly low priority. But yes, if for whatever reason, you cannot wait, I am ok with these changes as temporary changes.

But please note that the Broker and Cloud must have "identical" implementations when we are all set and done, so if the Cloud forces us into any type of spec, we will keep that same spec on the Broker. It is critical we use the identical models and API between Broker and Cloud

File: archfx_cloud/api/connection.py
Comment (dkarchmer) [reply archfx_cloud/api/connection.py:236]: > The difference goes farther than that. SimpleJWT requires using separate refresh token to refresh access token

I understand. I am simply debating the format of your payload returning the two tokens. You are using

```
"jwt": {
      "access": "access-token"
      "refresh": "refresh-token"
}
```

while I am saying that for backwards compatibility (yes, just for the normal login part), we may need to use

```
"jwt": "access-token",
"refresh-jwt": "refresh-token"
```

in both cases, we are returning two different tokens for the user to keep. One (the "access-token") that everybody needs, and the second ("refresh-token") that only people that need to call refresh need (only web app)

And no, most people will use `auth/login` as that is all that we document. `auth-jwt-login` is basically advanced/hidden 

> we can have best of both worlds. "jwt" will still be a string, but presense of "jwt-refresh" would signal SimpleJWT and different handling of the refresh it requires.

Yes, that is exactly what I was saying. It is just that I actually do like using the `dict` in the long run, as it is clean. So just was saying that I don't yet know if I can just change it and have the customer update there scripts (the good think is that I know their names) in the name of the greater long term good, or if I want to make it a little uglier but stay backwards compatible.

My only point is that you are going backwards here. It is not about the python package, but about what we want the feature to look like.

I know you want to just clean up the broker now, but there is no real reason why we need to do that right now. we can wait a few weeks until we decide on what the cloud should look like, and then just change the broker and python package to match that. 

Again, I am willing to be convinced that we should just approve this for now  so the broker can be cleaned up now, but either way, we will still need to do more work in the future:

1. Update the Broker to match whatever the final cloud implementation is
2. Clean up the python package once we no longer need to support two packages

Comment (dmaone) [reply archfx_cloud/api/connection.py:236]: I made the login backwards-compatible. It was pretty easy to do in retrospect :)

File: archfx_cloud/api/connection.py
Comment (dmaone) [reply archfx_cloud/api/connection.py:295]: It's not fully possible due to `MachineAPI` part of the Broker needing "admin" tokens (and not being in the cloud at all), but it can be made close enough. The change is about 20 lines of code. I'll put up a PR for that tomorrow, you'll see how easy it is :)

File: archfx_cloud/api/connection.py
Comment (dmaone) [reply archfx_cloud/api/connection.py:236]: https://github.com/iotile/archfx_broker/pull/1195 is the SimpleJWT swap - it's all in `server/`.

Tests are not actually changed that much - they just broken out into testing as close to a single thing as possible.

**Review by dkarchmer — APPROVED** [review]

**Pull request merged**
